File: OOP/04.EncapsulationAndPolymorphism/01.Shapes/Circle.cs
using System;
using Shapes.Interfaces;

namespace Shapes
{
    public class Circle : IShape
    {
        private const double PI = Math.PI;
        private const int MinRadius = 0;

        private double radius;

        public Circle(double radius)
        {
            this.Radius = radius;
        }

        public double Radius
        {
            get { return this.radius; }

            set
            {
                if (value <= MinRadius)
                {
                    throw new ArgumentOutOfRangeException("Radius cannot be negative or zero.");
                }
                this.radius = value;
            }
        }

        public double CalculateArea()
        {
            double area = PI * this.Radius * this.Radius;

            return area;
        }

        public double CalculatePerimeter()
        {
            double perimeter = 2 * PI * this.Radius;

            return perimeter;
        }

        public static double CalculateArea(double radius)
        {
            double area = PI * radius * radius;

            return area;

        }

        public static double CalculatePerimeter(double radius)
        {
            double perimeter = 2 * PI * radius;

            return perimeter;
        }
    }
}

File: JavaBasics/JavaOOP/Geometry/Shapes/PlaneShapes/Circle.java
package Geometry.Shapes.PlaneShapes;

import Geometry.Vertices.Vertex;
import com.sun.javaws.exceptions.InvalidArgumentException;

import java.util.ArrayList;

public class Circle extends PlaneShape {

    private Vertex center;
    private double radius;
    private ArrayList<Vertex> vertices;

    public Circle(Vertex center, double radius) throws InvalidArgumentException {
        super(center);
        this.center = super.getVertex();
        this.setRadius(radius);
        this.vertices = super.getVertices();
    }

    public double getRadius() {
        return this.radius;
    }

    public void setRadius(double radius) throws InvalidArgumentException {

        if (radius < 0) {
            String[] message = new String[]{"Radius cannot be negative."};
            throw new InvalidArgumentException(message);
        }

        this.radius = radius;
    }

    @Override
    public double getArea() {
        return Math.PI * radius * radius;
    }

    @Override
    public double getPerimeter() {
        return 2 * Math.PI * radius;
    }

    @Override
    public String toString() {
        StringBuilder circle = new StringBuilder();
        circle.append(String.format(
                "Circle:%nCenter = %s;%nRadius = %.2f;%nCircumference = %.2f;%nArea = %.2f;%n",
                this.center, this.radius, this.getPerimeter(), this.getArea()));

        return circle.toString();
    }
}
